test/utils/operationannotation/operationannotation.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2023 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operationannotation
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+)
+
+// Reconciler is a reconciler that removes operation annotations from objects.
+// This is useful for integration testing against an envtest control plane, which doesn't run the responsible
+// controller. Hence, if the tested controller must wait for the annotation to be removed, it will be stuck forever.
+type Reconciler struct {
+	Client    client.Client
+	ForObject func() client.Object
+}
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager) error {
+	if r.Client == nil {
+		r.Client = mgr.GetClient()
+	}
+
+	return builder.ControllerManagedBy(mgr).
+		Named("operationannotation").
+		For(r.ForObject(), builder.WithPredicates(r.hasOperationAnnotation())).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 5,
+			RecoverPanic:            true,
+		}).
+		Complete(r)
+}
+
+// Reconcile finalizes namespaces as soon as they are marked for deletion.
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+
+	obj := r.ForObject()
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	if obj.GetAnnotations()[v1beta1constants.GardenerOperation] == "" {
+		return reconcile.Result{}, nil
+	}
```

Comment:
Is this check need? predicate for this reconciler ensures that there will always be operation annotation on the object it reconciles.